src/sql/src/rbac.rs

```diff
@@ -325,410 +254,160 @@ pub fn is_rbac_enabled_for_session(system_vars: &SystemVars, session_vars: &Sess
-                ),
-                ObjectId::ClusterReplica((cluster_id, replica_id)) => {
-                    let cluster = catalog.get_cluster(cluster_id);
-                    let replica = catalog.get_cluster_replica(cluster_id, replica_id);
-                    let name = QualifiedReplica {
-                        cluster: cluster.name().into(),
-                        replica: replica.name().into(),
-                    };
-                    (ObjectType::ClusterReplica, name.to_string())
-                }
-                ObjectId::Database(id) => (
-                    ObjectType::Database,
-                    catalog.get_database(&id).name().to_string(),
-                ),
-                ObjectId::Schema((database_spec, schema_spec)) => {
-                    let schema = catalog.get_schema(&database_spec, &schema_spec);
-                    let name = catalog.resolve_full_schema_name(schema.name());
-                    (ObjectType::Schema, name.to_string())
-                }
-                ObjectId::Item(id) => {
-                    let item = catalog.get_item(&id);
-                    let name = catalog.resolve_full_name(item.name());
-                    (item.item_type().into(), name.to_string())
-                }
-                ObjectId::Role(_) => unreachable!("roles have no owner"),
-            })
-            .collect();
-        Err(UnauthorizedError::Ownership { objects })
-    } else {
-        Ok(())
-    }
-}
-
-/// Generates the privileges required to execute a given plan.
-///
-/// The result of this function is a set of tuples of the form
-/// (What object the privilege is on, What privilege is required, Who must possess the privilege).
-fn generate_required_privileges(
+/// Generates all requirements needed to execute a given plan.
+fn generate_rbac_requirements(
     catalog: &impl SessionCatalog,
     plan: &Plan,
+    active_conns: &BTreeMap<u32, RoleId>,
     target_cluster_id: Option<ClusterId>,
     resolved_ids: &ResolvedIds,
     role_id: RoleId,
-) -> Vec<(SystemObjectId, AclMode, RoleId)> {
-    // When adding a new plan, make sure that the plan struct is fully expanded. This helps ensure
-    // that when someone adds a field to a plan they get a compiler error and must consider any
-    // required changes to privileges.
+) -> RbacRequirements {
     match plan {
         Plan::CreateConnection(plan::CreateConnectionPlan {
             name,
             if_not_exists: _,
             connection: _,
             validate: _,
-        }) => {
-            vec![(
+        }) => RbacRequirements {
+            role_membership: BTreeSet::new(),
+            ownership: Vec::new(),
+            privileges: vec![(
                 SystemObjectId::Object(name.qualifiers.clone().into()),
                 AclMode::CREATE,
                 role_id,
-            )]
-        }
+            )],
+            item_usage: true,
+            superuser_action: None,
+        },
```

Comment:
Totally up to you, something that might reduce the boilerplate here is using the `..Default::default()` syntax, e.g.

```
impl Default for RbacRequirements {
  fn default() -> Self {
    RbacRequirements {
      role_membership: BTreeSet::default(),
      ownership: Vec::default(),
      privileges: Vec::default(),
      item_usage: true,
      superuser_action: None,
    }
  }
}

// ...

// When creating the struct, you only need to specify the non-default fields
RbacRequirements {
  privileges: vec![(
    SystemObjectId::Object(name.qualifiers.clone().into()),
    AclMode::CREATE,
    role_id,
  )],
  ..Default::default()
}
```

A tradeoff for reduced boilerplate is users wouldn't be forced to think about what fields to specify when adding a new Plan type

Reply:
I like it, I think it simplifies things. Just added.